### PR TITLE
[model:component] Add sampling techniques to address imbalanced dataset

### LINKED ISSUE #4281: [model:component] Add sampling techniques to address the imbalanced training dataset
The current training set does not utilize any sampling techniques (e.g., oversampling, undersampling, SMOTE) to address the imbalance in the dataset. We could implement appropriate sampling techniques to balance the dataset and equally represent bugs of all products and components.

### PR DESCRIPTION
**Pull request by benjaminmah**
Resolves #4281.

Investigating and adding sampling techniques (i.e. SMOTE, SMOTEEN, RandomUndersampling) to address the imbalanced dataset of bugs.

### CODE REVIEW

**Comment by benjaminmah** [comment]
Still a WIP. Metrics collected from SMOTE can be found here:  [metrics.log](https://github.com/user-attachments/files/16029517/metrics.log)

Takeaways: 
- SMOTE increases the training time to around 2 hours (as opposed to the current 30-40 minute training time)
- The precision and accuracy are extremely low

This is most likely due to the huge differences in the number of bugs in different products and components (1000+ vs 20), and SMOTE matches the number of bugs in each minority class to the majority class, making the ratio of synthetic data to real data very large.

**Comment by benjaminmah** [comment]
@suhaibmujahid Since this PR is mainly experimental, and the results were proven to be much worse, I think we can close this. WDYT?
